Features

gorazdko opened the pull request:
## Abstract

This PR supercedes #49 and additionally implements:
* sskr (encoding) as words and URs
* output descriptor and addresses for single native segwit wallet (base58 and ur in text and qr)
* refactored code
  * started work with menus with selected option highlighted

![IMG_20200921_004223](https://user-images.githubusercontent.com/25270775/93724106-87d03900-fba4-11ea-8ac2-6559af1a0dda.jpg)

I was able to restore the seed with bc-seedtool-cli from shards generated on lethekit:
```bash
# lethekit seed:
# e3b0c44298fc1c149afbf4c8996fb924
# ur:crypto-seed/oyadgdvlpfssfwmkzccebbnyztwkspnljlrhdkzcykahhf
# lethekit 2/5 SSKR shards
# tuna acid epic gyro keno bias able acid able puff back chef time jazz fund exam kiwi down jazz item belt gems cost undo hawk taco luck user tent
# tuna acid epic gyro keno bias able acid acid beta days lung mint part surf holy join scar huts need oboe exit dice high veto idle poem luck tent
# tuna acid epic gyro keno bias able acid also toys grim door hawk zone help view hill vast keno menu fund onyx jugs slot edge frog road user wand
# tuna acid epic gyro keno bias able acid apex iron knob ruin code exit user luck gift away king iced void taxi frog fuel many wasp need luck wand
# tuna acid epic gyro keno bias able acid aqua knob road lamb undo gyro jolt logo figs oval hard legs lion lion scar wand legs wasp skew echo calm
# ur:crypto-sskr/taadecgokobsaeadaepfbkcftejzfdemkidnjzimbtgsctuohktolkurtt

lethekit_seed=$(seedtool --in ur ur:crypto-seed/oyadgdvlpfssfwmkzccebbnyztwkspnljlrhdkzcykahhf)

expected_seed=$(seedtool --in sskr "tuna acid epic gyro keno bias able acid able puff back chef time jazz fund exam kiwi down jazz item belt gems cost undo hawk taco luck user tent" "tuna acid epic gyro keno bias able acid aqua knob road lamb undo gyro jolt logo figs oval hard legs lion lion scar wand legs wasp skew echo calm")

if [ $expected_seed == $lethekit_seed ];
then
    echo "Correct"
else
    echo "InCorrect"
fi
```


TODOs:
* bip39 is currently slow due to a function that collects all the words from scratch. These are needed for HD root key. I'll optimize it once i port from current bip39 lib to libwally bip39
* ~finish all menus with selected option highlighted~
* ~SSKR restore~
* ~network option is hidden atm~

ksedgwic commented:
I'm seeing the following when I build:
```
In file included from /home/user/bonsai/bc-lethekit/seedtool/seedtool.ino:5:0:
seed.h:11:25: fatal error: CborEncoder.h: No such file or directory
compilation terminated.
```

ksedgwic commented:
After resolving the `Library-Arduino-Cbor` dependency above I'm now seeing:
```
ResolveLibrary(avr/pgmspace.h)
  -> candidates: []In file included from /home/user/Arduino/libraries/bc-bip39/src/prefix1.h:5:0,
                 from /home/user/bonsai/bc-lethekit/seedtool/selftest.ino:11:
/home/user/Arduino/libraries/bc-bip39/src/arduino-support.h:5:26: fatal error: avr/pgmspace.h: No such file or directory
compilation terminated.
```

I believe this is a problem in the `bc-bip39` dependency, I'll follow up there ...

ksedgwic commented:
I found one bug, after multiple consecutive `Generate SSKR` traverals
the unit will hang.  Here is how to reproduce:

* Turn on, any key
* `A - Generate Seed`
* "123456", `#-Done`
* `#-Done` (back to `Seed Present` menu)
* `B -Generate SSKR`, 3, 5, `#-Next`
* `#-Next` 5 times
* `B -Generate SSKR`, 3, 5, `#-Next`
* `#-Next` 5 times
* `B -Generate SSKR`, 3, 5, `#-Next`
* `#-Next` 5 times
* `B -Generate SSKR`, 3, 5, `#-Next`

At this point it hangs, last thing in the log is:
```
15:15:44.890 -> tuna acid epic gyro news calm able also able apex dull ...
15:15:44.890 -> tuna acid epic gyro news calm able also acid leaf diet ...
15:15:44.890 -> tuna acid epic gyro news calm able also also news sets ...
15:15:44.890 -> tuna acid epic gyro news calm able also apex cola saga ...
15:15:44.890 -> tuna acid epic gyro news calm able also aqua away guru ...
```

Maybe this is a resource leak in SSKR generation?

I think ``SSKRShareSeq::shares_buffer` is leaked; I think a destructor for `SSKRShareSeq` should be created which frees any existing buffer memory.  Maybe that will fix this problem.

ksedgwic commented:
General Observations:

* Everything worked with both display types. The consecutive SSKR generation bug existed with both display types.

* Some trouble with clipping of the "text" (non-QR) output displays. Alignment of the physical dispay in the LetheKit is unlikely to be very precise; wonder if we can find a solution/workaround.

* The `Enter derivation path` submenu of `Set derivation path` worked quite well.
  
Possible (Future?) Improvements:

* Should there be a "original/legacy" (P2PKH) choice in `Set derivation path`?

* Is there any way we could fit the labels `P2WPKH` and `P2SH-P2WPKH` associated with the choices in the `Set derivation path` menu?  They are very specific and useful for experts.
  
* `Open Wallet` -> `show address` appears to show P2WPKH addresses; should it be able to show `P2SH-P2WPKH` (and `P2PKH`) as well?

I bet I missed a memo on these:

* Is `Restore SSKR` no longer supported?

* Looks like the `network` setting is hidden in the UI; is `REGTEST` the best default in this case?  Maybe this is to prevent use on mainnet?

gorazdko commented:
Thanks @ksedgwic for testing. These are all valid points.

I think for this PR besides resolving bugs I will also push the SSKR restore feature and finish the UI refactoring. Other points will be part of the follow up PRs.

gorazdko commented:
@ksedgwic I have now implemented sskr_restore and fixed memory leaks

I have also refactored the gui with highlighted options

Can you re-review?

gorazdko commented:
Also note: wrt `sskr restore`, you won't be able to input the word `zest` if it happens to be in the shard. This is due to a bug in `bytewords` which is tracked here https://github.com/BlockchainCommons/Research/issues/45

gorazdko commented:
> Also note: wrt sskr restore, you won't be able to input the word zest if it happens to be in the shard. This is due to a bug in bytewords which is tracked here BlockchainCommons/Research#45

Now fixed.

Also, merging this PR will require https://github.com/BlockchainCommons/bc-bytewords/pull/6

ksedgwic commented:
Awesome!  I'll try and find time this week.

On Mon, Oct 5, 2020 at 3:15 PM Gorazd Kovacic <notifications@github.com>
wrote:

> Also note: wrt sskr restore, you won't be able to input the word zest if
> it happens to be in the shard. This is due to a bug in bytewords which is
> tracked here BlockchainCommons/Research#45
> <https://github.com/BlockchainCommons/Research/issues/45>
>
> Now fixed.
>
> Also, merging this PR will require BlockchainCommons/bc-bytewords#6
> <https://github.com/BlockchainCommons/bc-bytewords/pull/6>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlockchainCommons/bc-lethekit/pull/50#issuecomment-703918059>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAJNPXAGO6QEYVC3PYVTGDSJJAO7ANCNFSM4RT4226Q>
> .
>


-- 
Ken Sedgwick
Bonsai Software, Inc.
http://www.bonsai.com/ken/
(510) 269-7334
ken@bonsai.com
Public Key: http://www.bonsai.com/ken/ken.asc
GPG Fingerprint: 4695 E5B8 F781 BF85 4326  9639 BBFC E515 8602 5550

ksedgwic commented:
I'm seeing the following error when building/uploading:
```
In file included from /tmp/arduino_build_985968/sketch/keystore.h:4:0,
                 from /tmp/arduino_build_985968/sketch/userinterface.h:7,
                 from /home/user/bonsai/bc-lethekit/seedtool/seedtool.ino:6:
/home/user/Arduino/libraries/secp256k1-embedded/src/secp256k1.h:2:44: fatal error: ../secp256k1/include/secp256k1.h: No such file or directory
 #include "../secp256k1/include/secp256k1.h"
                                            ^
```

My `secp256k1-embedded` dir looks like this:
```
.
├── LICENSE
├── README.md
├── examples
│   ├── basic_example
│   │   └── basic_example.ino
│   ├── mbed
│   │   ├── main.cpp
│   │   └── mbed_app.json
│   └── secp256k1.py
├── keywords.txt
├── library.properties
├── micropython.mk
├── mpy
│   ├── config
│   │   ├── ecmult_static_context.h
│   │   ├── ext_callbacks.c
│   │   └── libsecp256k1-config.h
│   └── libsecp256k1.c
├── secp256k1
├── secp256k1.lib
└── src
    ├── ecmult_const_impl.h
    ├── ecmult_static_context.h
    ├── include
    │   ├── secp256k1.h
    │   ├── secp256k1_ecdh.h
    │   ├── secp256k1_preallocated.h
    │   └── secp256k1_recovery.h
    ├── libsecp256k1-config.h
    ├── secp256k1.h
    ├── secp256k1_bundle.c
    ├── secp256k1_ecdh.h
    ├── secp256k1_preallocated.h
    └── secp256k1_recovery.h
```
Note the `secp256k1` subdir is empty.

Please advise.

gorazdko commented:
secp256k1 must not be empty.

Submodules must be pulled recursively:

```
git submodule update --init --recursive
export LK_ROOT=`pwd` && export ARDUINO_ROOT=~/Arduino && $LK_ROOT/scripts/install-lethekit $LK_ROOT $ARDUINO_ROOT
```

ksedgwic commented:
Bah, I think I missed that a previous time ... I made a procedure now.  Apologies.

Now I'm seeing:
```
In file included from /home/user/bonsai/bc-lethekit/seedtool/selftest.ino:17:0:
test_bc_ur.hpp:5:24: fatal error: ArduinoSTL.h: No such file or directory
 #include <ArduinoSTL.h>
                        ^
```
Thoughts?

gorazdko commented:
> Using the managed ArduinoSTL package I'm seeing:

Fixed now. The package manager has an older version. It has to be installed from github.

ChristopherA commented:
Again, I need at least someone other than @gorazdko to confirm this version before I can merge it. @ksedgwic ? Do we know of any others? I know @wolfmcnally has installed an update to his #LetheKit at least once.

ksedgwic commented:
Builds fine for me now.

Runs well with both display types.

The "multiple consecutive Generate SSKR" bug appears to be fixed, couldn't reproduce it even with many generates.

I was able to generate SSKR shares and recover with a set of them.

I think it's looking good.

ksedgwic commented:
ACK+

ChristopherA reviewed (approved):
I see ACK from @ksedgwic of this PR from @gorazdko, which is sufficient to merge this PR to master. Did not test it myself.
